ROADMAP.md

```diff
@@ -1,70 +1,69 @@
+> cert-manager is the easiest way to automatically manage certificates in Kubernetes clusters.
 
-cert-manager should be able to deliver and manage X.509 certificates to popular
-projects in the cloud-native ecosystem.
+This applies to cert-manager itself, and to other subprojects which are part of the wider cert-manager project
+including:
 
-- Service Mesh Integration: While we have good Istio and Open Service Mesh integration, expand to other projects such as Linkerd, cilium
+- trust-manager
+- csi-driver
+- csi-driver-spiffe
+- istio-csr
+- approver-policy
 
-### Adoption of upstream APIs
+## Roadmap Items
 
-Continue to support latest APIs for upstream K8s and related SIGs.
+### Adoption of Upstream Changes
 
-- Kubernetes APIs: keep up to date with Kubernetes API changes and release cadence
-- CSR API: support the sig-auth CSR API for certificate requests in kubernetes
-- [Trust Anchor Sets](https://github.com/kubernetes/enhancements/pull/3258)
-- Gateway API
+Continue to support latest versions, new APIs for upstream Kubernetes and related upstream projects.
 
-### Extensibility
+For example:
 
-Widen the scope of integrations with cert-manager.
+- Kubernetes APIs: keep up to date with Kubernetes API changes
+- Kubernetes Versions: maintain support for a reasonable spread of Kubernetes versions, and support new versions as quickly as we can
+- [ClusterTrustBundle](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.29/#clustertrustbundle-v1alpha1-certificates-k8s-io) support
+- Gateway API
 
-- EST support: support a standard for ACME-like issuance within an enterprise
-- External DNS plugin: enable ACME DNS01 requests to be completed using external-dns
-- Improve external issuer development experience: documentation and examples for people developing external issuers
+#### Goals for cert-manager `v1.15`
 
-### PKI lifecycle
+- [x] Graduate Gateway API support to Beta ([#6961](https://github.com/cert-manager/cert-manager/pull/6961))
+- [ ] Support Kubernetes `v1.30` ([#7006](https://github.com/cert-manager/cert-manager/issues/7006))
 
-Enable best-practice PKI management with cert-manager.
+### Shrinking Core / Aiding Extensibility
 
-- Handle CA certs being renewed: deal with the cases where the CA cert is renewed and allow for all signed certs to be renewed
-- Make cert-manager a viable way to create and manage private PKI deployments at scale
-- Trust root distribution: handle distributing all trust roots within a cluster, solving trust for private and public certificates
+Minimise the surface area of cert-manager's core components to reduce attack surface, binary size, container size and complexity
 
-See also [trust-manager](https://cert-manager.io/docs/projects/trust/) for more on trust distribution.
+For example:
 
-### End-user experience
+- Move "core" issuers with dependencies (ACME, Vault, Venafi) into external issuers, which might still be bundled by default
```

Comment:
By "external issuers", do you mean issuers that are still in "core" (i.e., in the the same codebase, github.com/cert-manager/cert-manager) but in different go.mod and binaries/containers?

Or do you mean "external issuers" as a separate GitHub project and separate Helm chart?

Reply:
It would be good to clarify this a bit. From an engineering perspective, I understand the aim of breaking things down into smaller pieces. However, as an end-user, if we threat everything as "external issuers," it would mean extra configuration work (extra installations) and extra containers to run (more resources).

Reply:
Both very fair comments, thank you both!

When this has been discussed in the past we've tended to take the line that since we hate breaking changes, "splitting things out of core" would mean that if you continue to install cert-manager in the same way (e.g. from the same named helm chart in the same location) you wouldn't see a difference at all.

But under the hood, all of the issuers would be running in a separate pod, allowing users to opt out of running any of the core issuers if they wanted to, say, only run the AWS PrivateCA issuer.

Current "core" DNS webhooks would probably end up included with the ACME issuer in that model.

There would be a slight overhead from having to run "cert-manager-core-issuers" pods as well as the "cert-manager-controller" pod but the cost/benefit analysis in my mind has always been that that'd be well worth it.

There are security benefits to this too (more tightly scoped RBAC meaning a popped pod can't do as much), and obvious maintainence benefits from spreading dependencies across different modules.

I think what I wrote there is probably too specific for a roadmap. There's an old [design doc](https://github.com/cert-manager/cert-manager/pull/5848/files) I wrote which didn't go anywhere - maybe I could update that doc and link to it?

Reply:
I've updated to make it clear that nothing in this section should break anyone. For now, I think that should be clear enough since nobody is necc. going to be working on this in the immediate future. Thanks both!